pandas/core/index.py

```diff
@@ -1416,7 +1416,11 @@ def shift(self, periods=1, freq=None):
             # OK because immutable
             return self
 
-        offset = periods * freq
+        try:
+            offset = periods * freq
+        except TypeError:
+            raise NotImplementedError("freq should be defined for %s" % self.dtype)
+
```

Comment:
simply raise ``NotImplementedError`` here always; define this function instead in ``tseries/base/DatetimeIndexOpsMixin`` (w/o the ``NotImplemenetdError``)